### ISSUE TITLE
Execution time anomaly

### ISSUE THREAD

**Issue by mokun** [mars-sim/mars-sim#1204]
**Describe the bug**
- On my 7th gen intel i7 machine with 8 cores, the 15-settlement default scenario template has an average execution time of 10-20 ms per frame.
- Note that the average execution time would vary from machine to machine.
- However, recent changes has caused anomalous spike of execution time as high as 5000 ms, thus more than 300x the normal amount of time it takes to complete one frame

**Stated Goal(s)**
- Benchmark the execution time within each category of Settlement's, Person's, and/or Building's timePassing() 
- Identify any unusual time loop that exceed the normal average.

**Discuss any potential problems you have and have not ruled out**
- Currently, other threads do not use that much CPU time, as shown in VisualVM.
 
**Stacktrace**
```
2024-01-03T08:35:35.0 (Config) MarsProject : Starting Mars Simulation Project 3.7.0 - Build 09e05a7-dirty - 64-bit Java 21.0.1 - 8 Cores
2024-01-03T08:35:35.0 (Config) MarsProject : List of input args : [-new, -sandbox, -timeratio, 576]
01-Adir-01:000.928 (Config) MasterClock : sleep=-380 ms, desiredTR=576, actualTR=526.54, millisol/sec=6.54, pulse/sec=14.11, mpp=70.88, ms/pulse=70.88, execution=450 ms
01-Adir-01:001.394 (Config) MarsProject : Starting the Main Window...
01-Adir-01:001.394 (Config) MainWindow : Running mars-sim in Sandbox Mode.
01-Adir-01:001.394 (Config) MasterClock : sleep=-4 ms, desiredTR=576, actualTR=483.07, millisol/sec=6.59, pulse/sec=14.19, mpp=70.48, ms/pulse=70.48, execution=74 ms
01-Adir-01:001.862 (Info) AlgaeFarming : [Tian Cheng] Algae Pond 1 - Algae fully tended for 0.04 millisols.
01-Adir-01:001.862 (Config) MasterClock : sleep=28 ms, desiredTR=576, actualTR=490.09, millisol/sec=6.59, pulse/sec=14.16, mpp=70.60, ms/pulse=70.60, execution=42 ms
01-Adir-01:002.330 (Config) MasterClock : sleep=47 ms, desiredTR=576, actualTR=498.07, millisol/sec=6.58, pulse/sec=14.13, mpp=70.75, ms/pulse=70.75, execution=23 ms
01-Adir-01:002.800 (Config) MasterClock : sleep=33 ms, desiredTR=576, actualTR=505.39, millisol/sec=6.57, pulse/sec=14.11, mpp=70.89, ms/pulse=70.89, execution=37 ms
01-Adir-01:003.271 (Config) MasterClock : sleep=39 ms, desiredTR=576, actualTR=514.54, millisol/sec=6.56, pulse/sec=14.08, mpp=71.04, ms/pulse=71.04, execution=32 ms
01-Adir-01:003.271 (Config) MainWindow : 2 screens detected.
01-Adir-01:003.271 (Config) MainWindow : The last saved window dimension is 1550 x 774.
01-Adir-01:003.271 (Config) MainWindow : The last saved frame starts at (-2, 2).
01-Adir-01:003.742 (Config) MasterClock : sleep=59 ms, desiredTR=576, actualTR=520.42, millisol/sec=6.55, pulse/sec=14.05, mpp=71.15, ms/pulse=71.15, execution=12 ms
01-Adir-01:004.214 (Config) MasterClock : sleep=59 ms, desiredTR=576, actualTR=525.03, millisol/sec=6.55, pulse/sec=14.04, mpp=71.25, ms/pulse=71.25, execution=12 ms
01-Adir-01:004.687 (Config) MasterClock : sleep=59 ms, desiredTR=576, actualTR=530.84, millisol/sec=6.54, pulse/sec=14.02, mpp=71.35, ms/pulse=71.35, execution=12 ms
01-Adir-01:004.687 (Config) MainDesktopPane : The main desktop pane is initially set to 1550 x 774.
01-Adir-01:005.161 (Config) MasterClock : sleep=50 ms, desiredTR=576, actualTR=535.35, millisol/sec=6.53, pulse/sec=14.00, mpp=71.43, ms/pulse=71.43, execution=21 ms
01-Adir-01:005.635 (Config) MasterClock : sleep=53 ms, desiredTR=576, actualTR=530.76, millisol/sec=6.54, pulse/sec=14.00, mpp=71.40, ms/pulse=71.40, execution=18 ms
01-Adir-01:006.110 (Config) MasterClock : sleep=57 ms, desiredTR=576, actualTR=536.22, millisol/sec=6.53, pulse/sec=13.99, mpp=71.50, ms/pulse=71.50, execution=14 ms
01-Adir-01:006.585 (Info) Fishery : [Terminus] Large Greenhouse 1 - Weeds fully tended for 2.0 millisols.
01-Adir-01:006.585 (Info) Fishery : [New Seoul] Fish Farm 1 - Weeds fully tended for 2.0 millisols.
01-Adir-01:006.585 (Config) MasterClock : sleep=42 ms, desiredTR=576, actualTR=540.41, millisol/sec=6.53, pulse/sec=13.97, mpp=71.58, ms/pulse=71.58, execution=29 ms
01-Adir-01:007.061 (Config) MasterClock : sleep=39 ms, desiredTR=576, actualTR=545.04, millisol/sec=6.52, pulse/sec=13.96, mpp=71.65, ms/pulse=71.65, execution=32 ms
01-Adir-01:007.537 (Config) MasterClock : sleep=61 ms, desiredTR=576, actualTR=549.25, millisol/sec=6.52, pulse/sec=13.94, mpp=71.73, ms/pulse=71.73, execution=10 ms
01-Adir-01:008.014 (Config) MasterClock : sleep=67 ms, desiredTR=576, actualTR=551.51, millisol/sec=6.52, pulse/sec=13.93, mpp=71.77, ms/pulse=71.77, execution=4 ms
01-Adir-01:008.491 (Config) MasterClock : sleep=60 ms, desiredTR=576, actualTR=554.36, millisol/sec=6.51, pulse/sec=13.92, mpp=71.82, ms/pulse=71.82, execution=11 ms
01-Adir-01:008.968 (Config) MasterClock : sleep=64 ms, desiredTR=576, actualTR=549.36, millisol/sec=6.52, pulse/sec=13.93, mpp=71.77, ms/pulse=71.77, execution=7 ms
01-Adir-01:009.446 (Info) TendGreenhouse : [Dream City - Inflatable Greenhouse 2] GardenBot 002 - Cleaning curtains.
01-Adir-01:009.446 (Config) MasterClock : sleep=63 ms, desiredTR=576, actualTR=551.73, millisol/sec=6.52, pulse/sec=13.92, mpp=71.82, ms/pulse=71.82, execution=8 ms
01-Adir-01:009.924 (Info) Fishery : [Heritage] Central Hub A 1 - Weeds fully tended for 2.0 millisols.
01-Adir-01:009.924 (Info) TendGreenhouse : [Heritage - Large Greenhouse 1] Dewey #1 - Cleaning curtains.
01-Adir-01:009.924 (Config) MasterClock : sleep=66 ms, desiredTR=576, actualTR=555.49, millisol/sec=6.51, pulse/sec=13.91, mpp=71.88, ms/pulse=71.88, execution=5 ms
01-Adir-01:010.402 (Warning) Storage : [Hong Tian] Ran out of Strawberry.
01-Adir-01:010.402 (Config) MasterClock : sleep=-61 ms, desiredTR=576, actualTR=558.10, millisol/sec=6.51, pulse/sec=13.90, mpp=71.93, ms/pulse=71.93, execution=132 ms
01-Adir-01:010.881 (Info) TendGreenhouse : [Heritage - Large Greenhouse 1] Manny Kung - Inspecting irrigation.
01-Adir-01:010.881 (Config) MasterClock : sleep=62 ms, desiredTR=576, actualTR=534.24, millisol/sec=6.54, pulse/sec=13.96, mpp=71.65, ms/pulse=71.65, execution=9 ms
01-Adir-01:011.360 (Info) TendGreenhouse : [Heritage - Fish Farm 1] GardenBot 018 - Cleaning curtains.
01-Adir-01:011.360 (Config) MasterClock : sleep=6 ms, desiredTR=576, actualTR=537.53, millisol/sec=6.53, pulse/sec=13.94, mpp=71.71, ms/pulse=71.71, execution=65 ms
01-Adir-01:011.839 (Config) MasterClock : sleep=63 ms, desiredTR=576, actualTR=542.92, millisol/sec=6.53, pulse/sec=13.93, mpp=71.80, ms/pulse=71.80, execution=8 ms

```

**Specifications**
- post 3.7.0 build 9186

**Additional context**
- We have been using this [thead](https://github.com/mars-sim/mars-sim/issues/655) to handle this issue for a while recently.

**Comment by mokun** [comment]
@bevans2000,

In `EVAOperation`, we have

```java

	/**
	 * Aborts an EVA, if the Person is outside get them to return otherwise end the Task.
	 * 
	 * @param reason Reason for ending.
	 */
	protected void abortEVA(String reason) {
		if (reason != null) {
 			logger.info(worker, "EVA " + getName() + " aborted: " + reason);
		}
		
		if (person.isOutside()) {
            setPhase(WALK_BACK_INSIDE);
		}
    	else
        	endTask();
	}
```

The only instance where the abortEVA("End of Work Shift.") is called is found in `Diglocal`, 
```java
    /**
     * Drops off resources from container.
     * 
     * @param time
     * @return
     */
	private double dropOffResource(double time) {
    	double remainingTime = time;
    	
    	Container container = person.findContainer(containerType, false, resourceID);
    	if (container == null)
    		return 0D;
    	
    	double amount = container.getAmountResourceStored(resourceID);	
    	
        if (amount > 0) {
        	 	
           	double portion = LOADING_RATE * time;
           	         	
           	if (portion > amount) {
           		portion = amount;
           	} 	
           	
            if (portion > 0) {
	        	double loadingTime = portion / LOADING_RATE;
	        	
	        	remainingTime = remainingTime - loadingTime;
	        	
				// Transfer the resource out of the Container
				unloadContainer(container, portion, time);
	            
	        	if (isDone() || getTimeLeft() - time < 0) {
	    			// Need to extend the duration so as to drop off the resources
	    			setDuration(getDuration() + EXTENDED_TIME);
	
	    			return remainingTime;
	    		}
            }
        }
        else if (!person.isOnDuty()) {
			// Duty has ended so abort digging
			abortEVA("End of Work Shift.");
		}
		else {
        	// Reset this holder
        	collectionLimit = 0;
        	// Go back to the digging site
        	setPhase(WALK_TO_OUTSIDE_SITE);
        }
        
    	return remainingTime;
    }
```

**Comment by mokun** [comment]
If I focus on the benchmark on Person and WalkOutside, it's very clear that WalkOutside is requiring ***normally high*** execution time (between 10 and 35 ms)
```
01-Adir-02:058.133 (Severe) WalkOutside : [22.4000 N 18.0000 W] Lara Klein - elapsedMS: 19
01-Adir-02:058.133 (Severe) Person : [22.4000 N 18.0000 W] Lara Klein - elapsedMS: 19
01-Adir-02:060.809 (Severe) WalkOutside : [22.4000 N 18.0000 W] Benicio Ramos - elapsedMS: 15
01-Adir-02:060.809 (Severe) Person : [22.4000 N 18.0000 W] Benicio Ramos - elapsedMS: 15
...
01-Adir-02:071.988 (Severe) WalkOutside : [22.4000 N 18.0000 W] Zsófia Németh - elapsedMS: 15
01-Adir-02:071.988 (Severe) Person : [22.4000 N 18.0000 W] Zsófia Németh - elapsedMS: 15
01-Adir-02:071.988 (Info) PlanMission : [Hartmann Base] Emily Reuter - Submitted a mission plan for Regolith Prospecting 2-001.
01-Adir-02:071.988 (Info) MissionManager : [Hartmann Base] Emily Reuter - Put together a mission plan for Regolith Prospecting 2-001.
01-Adir-02:073.336 (Severe) WalkOutside : [22.4000 N 18.0000 W] William Christensen - elapsedMS: 13
01-Adir-02:073.336 (Severe) Person : [22.4000 N 18.0000 W] William Christensen - elapsedMS: 13
01-Adir-02:075.583 (Severe) WalkOutside : [22.4000 N 18.0000 W] Benicio Ramos - elapsedMS: 15
01-Adir-02:075.583 (Severe) Person : [22.4000 N 18.0000 W] Benicio Ramos - elapsedMS: 15
01-Adir-02:076.032 (Severe) WalkOutside : [22.4000 N 18.0000 W] Lara Klein - elapsedMS: 13
01-Adir-02:076.032 (Severe) Person : [22.4000 N 18.0000 W] Lara Klein - elapsedMS: 13
01-Adir-02:076.931 (Severe) WalkOutside : [22.4000 N 18.0000 W] Zsófia Németh - elapsedMS: 14
01-Adir-02:076.931 (Severe) Person : [22.4000 N 18.0000 W] Zsófia Németh - elapsedMS: 14
01-Adir-02:077.830 (Severe) WalkOutside : [22.4000 N 18.0000 W] Elias Meier - elapsedMS: 17
01-Adir-02:077.830 (Severe) Person : [22.4000 N 18.0000 W] Elias Meier - elapsedMS: 17
01-Adir-02:079.627 (Severe) WalkOutside : [22.4000 N 18.0000 W] Haru Yamazaki - elapsedMS: 12
01-Adir-02:079.627 (Severe) Person : [22.4000 N 18.0000 W] Haru Yamazaki - elapsedMS: 13
01-Adir-02:081.873 (Severe) WalkOutside [x2] : [22.4000 N 18.0000 W] Zsófia Németh - elapsedMS: 19
01-Adir-02:081.873 (Severe) Person [x2] : [22.4000 N 18.0000 W] Zsófia Németh - elapsedMS: 19
01-Adir-02:081.873 (Info) Charge [x3] : [Hartmann Base] ChefBot 002 - Switching to wireless charging in Central Hub A 1.
01-Adir-02:082.321 (Severe) WalkOutside : [22.4000 N 18.0000 W] Rakesh Nirmal - elapsedMS: 11
01-Adir-02:082.321 (Severe) Person : [22.4000 N 18.0000 W] Rakesh Nirmal - elapsedMS: 11
01-Adir-02:082.770 (Severe) WalkOutside : [22.4000 N 18.0000 W] Anna Bianchi - elapsedMS: 35
01-Adir-02:082.770 (Severe) Person : [22.4000 N 18.0000 W] Anna Bianchi - elapsedMS: 35
```

But the worst part is that once in a while we have unusually high spike that eclipses everything else : 
```java
01-Adir-02:062.147 (Severe) WalkOutside : [22.4000 N 18.0000 W] Noah Mäkelä - elapsedMS: 4857
01-Adir-02:062.147 (Severe) Person : [22.4000 N 18.0000 W] Noah Mäkelä - elapsedMS: 4861
01-Adir-02:062.147 (Config) MasterClock : execution time = 4873 ms
```

**Comment by mokun** [comment]
@bevans2000 

Does the conversion from meter to centimeter affect WalkOutside's determineWalkingPath() ? 

At least we know this method is now using a lot of CPU resources.

**Comment by mokun** [comment]
The biggest problem seems to be in `LocalPosition`'s equals() as follows :
```java
	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		if (getClass() != obj.getClass())
			return false;
		LocalPosition other = (LocalPosition) obj;
//		double xdiff = x - other.x;
//		if (Math.abs(xdiff) > ONE_CENTIMETER)
//			return false;
//		double ydiff = y - other.y;
//		return (Math.abs(ydiff) < ONE_CENTIMETER);
		if (Double.doubleToLongBits(x) != Double.doubleToLongBits(other.x))
			return false;
		if (Double.doubleToLongBits(y) != Double.doubleToLongBits(other.y))
			return false;
		return true;
	}
```
If we use `Double.doubleToLongBits()`, it no longer causes execution time to go beyond 70 ms due to WalkOutside.

Why is java's `Math.abs()` so bad ?

**Comment by mokun** [comment]
Oh it still creates spikes once in a long while, even though much less now (say it occurs once a day or once every two days). 

Before the 3000 ms spike was ~4 times a day on average.

```
01-Adir-02:138.118 (Severe) WalkOutside : [22.4000 N 18.0000 W] Ryuto Kaneko - elapsedMS: 4338
01-Adir-02:138.118 (Severe) Person : [22.4000 N 18.0000 W] Ryuto Kaneko - elapsedMS: 4338
01-Adir-02:138.118 (Severe) Settlement : [Hartmann Base] elapsedMS: 4342
01-Adir-02:138.118 (Config) MasterClock : execution time = 4342 ms
```

**Comment by mokun** [comment]
CORRECTION: Now that after I test run a dozen times, it turns out using either methods in equals() have the same probability of getting high execution time spike. No difference made. :(

**Comment by bevans2000** [comment]
I would comment out the aborting at the end of the shift to see if that has an impact. The message is actually a sign that a shift has changed and it might be the changing shift that is causing a problem.
However this has not been a problem before.
I think the problem will be something looping with a small time increment.

**Comment by bevans2000** [comment]
Make sure you run the Unit tests because the Local Position equals method might not work when using longboat because it is an estimate not exact.

**Comment by mokun** [comment]
>  comment out the aborting at the end of the shift to see if that has an impact.

I already tried commenting it out. No impact. WalkOutside is still causing time spikes.

**Comment by mokun** [comment]
> the Local Position equals method might not work when using longboat because it is an estimate not exact.

I've gone back to using the range comparison, instead of using Double.doubleToLongBits(x) : 
```java
		double xdiff = x - other.x;
		if (Math.abs(xdiff) > ONE_CENTIMETER)
			return false;
		double ydiff = y - other.y;
		return (Math.abs(ydiff) < ONE_CENTIMETER);
```

**Comment by mokun** [comment]
The situation with the time spike is mostly solved in #1207. Closing this.
